bdb-ip.md

```diff
@@ -0,0 +1,437 @@
+  /transactions/ID returns non-committed (`UNDECIDED`, `BACKLOG`) transactions.
+- Ideally, the inner workings of BDB do not need to be fully understood in
+  order to work with BDB (compare to databases: Many database users aren't
+  experts on the details of the database's inner workings: "What is a BACKLOG,
+  and why should I care?")
+
+
+##### Proposal 2
+
+- /transaction/ID is returning only transactions included in a VALID block
+
+
+Pros:
+
+- Expected behavior (similar to comparable systems: databases and blockchains
+alike)
+- Transaction's status can still be monitored at /status, as is expected
+- Non-persisted transaction cannot be queried from BDB
+- No mental overhead
+- Interface is safe: Users will be making mission-critical assumptions on the
+  /transactions/id endpoint but might miss the API docs part. We want it's
+  default behavior to be as safe as possible
+- Querying for UNDECIDED or BACKLOG is very difficult to achieve without
+  confusing the user (e.g. [transactions can be in multiple
+  states](https://github.com/bigchaindb/bigchaindb/issues/1038#issuecomment-274049790)).
+  Safes us and the user lots of time
+
+Contra:
+
+- Transactions from BACKLOG OR UNDECIDED blocks cannot be queried anymore
+(actually they can using /blocks and /votes)
+
+
+##### Proposal 3
+
+- /transaction/ID is returning only transactions included in a VALID block
+- /transaction/ID?safe=false (where ?safe=true or lack of is default) may
+return transactions from BAcKLOG or UNDECIDED blocks
+
+
+Pros:
+
+- Expected behavior (similar to comparable systems: databases and blockchains
+alike)
+- Non-persisted transaction can be queried from BDB (is that a Pro?)
+- Interface is safe: Users will be making mission-critical assumptions on the
+  /transactions/id endpoint but might miss the API docs part. We want it's
+  default behavior to be as safe as possible
+
+Contra:
+
+- Might be used as a substitute endpoint instead of /status?tx_id
+- Mental overhead
+- Ideally, the inner workings of BDB do not need to be fully understood in
+  order to work with BDB (compare to databases: Many database users aren't
+  experts on the details of the database's inner workings: "What is a BACKLOG,
+  and why should I care?")
+
+
+#### Proposal 4
+
+- /transactions/id gets a ?status flag
+
+Contra:
+
+- Querying for UNDECIDED or BACKLOG is very difficult to achieve without
+  confusing the user (e.g. [transactions can be in multiple
+  states](https://github.com/bigchaindb/bigchaindb/issues/1038#issuecomment-274049790))
+  Safes us and the user lots of time
+- Might be used as a substitute endpoint instead of /status?tx_id
+- Ideally, the inner workings of BDB do not need to be fully understood in
+  order to work with BDB (compare to databases: Many database users aren't
+  experts on the details of the database's inner workings: "What is a BACKLOG,
+  and why should I care?")
+
+
+##### Personal favorite: Proposal 2
```

Comment:
There is one problem with Proposal 2.

1. The user does a POST and gets a `202`
2. The user does a GET for that transaction id and gets a `404` (the transaction is still in the BACKLOG or UNDECIDED but a this point the user has no idea what happened)
3. The user does another GET for that transaction id and now it gets a `200`

Another proposal would be mix of proposal 1 and proposal 4 where instead of using a status flag the transaction document always contains the status:

```json
{
    "transaction": {"..."},
    "status": "VALID | UNDECIDED | BACKLOG"
}
```
but this still suffers from all the contras of Proposal 4.

Or make use of HTTP status codes:
```python
response = get('transactions/04c00267af82c161b4bf2ad4a47d1ddbfeb47eef1a14b8d51f37d6ee00ea5cdd')
if response.status_code == 202:
    # transaction still being processed
elif response.status_code == 201:
    # transaction is valid and was persisted
elif response.status == 404:
    # transaction was not persisted
```